src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+//! ```
+//! (note that the ranges on the LHS and RHS of `"mult with 2^(2n-2)"` are stated
+//!  here for vectors with a norm less than `1`)
+//!
+//! Given a vector `ys` of numbers in the field integer encoding (in `[0,2^n)`),
+//! this gives the following equation:
+//! ```text
+//! our_norm_on_encoded(ys) = our_norm([dec(y) for y in ys])
+//!                         = 2^(2n-2) * sum_{y in ys} ((y - 2^(n-1)) * 2^(1-n))^2
+//!                         = 2^(2n-2)
+//!                           * sum_{y in ys} y^2 - 2*y*2^(n-1) + (2^(n-1))^2
+//!                           * 2^(1-n)^2
+//!                         = sum_{y in ys} y^2 - (2^n)*y + 2^(2n-2)
+//! ```
+//!
+//! Let `d` denote the number of the vector entries. The maximal value the result
+//! of `our_norm_on_encoded()` can take occurs in the case where all entries are
+//! `2^n-1`, in which case `d * 2^(2n-2)` is an upper bound to the result. The
+//! finite field used for encoding must be at least as large as this.
+//! For validating that the norm of the submitted vector lies in the correct
+//! range, consider the following:
+//!  - The result of `norm(xs)` should be in `[0,1)`.
+//!  - Thus, the result of `our_norm(xs)` should be in `[0,2^(2n-2))`.
+//!  - The result of `our_norm_on_encoded(ys)` should be in `[0,2^(2n-2))`.
+//! This means that the valid norms are exactly those representable with `2n-2`
+//! bits.
+//!
+//! ### Differences in the computation because of distribution
+//!
+//! In `decode_result()`, what is decoded are not the submitted entries of a
+//! single client, but the sum of the the entries of all clients. We have to
+//! take this into account, and cannot directly use the `dec()` function from
+//! above. Instead we use:
+//! ```text
+//! dec'(x) = d * 2^(1-n) - c
+//! ```
+//! Here, `c` is the number of clients.
+//!
+//! ### Naming in the implementation
+//!
+//! The following names are used:
+//!  - `self.bits_per_entry` is `n`
+//!  - `self.entries`        is `d`
+//!  - `self.bits_for_norm`  is `2n-2`
+//!
+//! ### Submission layout
+//!
+//! The client submissions contain a share of their vector and the norm
+//! they claim it has.
+//! The submission is a vector of field elements laid out as follows:
+//! ```text
+//! |---- bits_per_entry * entries ----|---- bits_for_norm ----|
+//!  ^                                  ^
+//!  \- the input vector entries        |
+//!                                     \- the encoded norm
+//! ```
+//!
+//! ### Validity
+//!
+//! In addition to checking that every submission entry is `0` or `1`, the validation
+//! circuit of this type computes the norm and compares to what the client
+//! claimed.
+//!
+//! ### Value `1`
+//!
+//! We actually do not allow the submitted norm or vector entries to be
+//! exactly `1`, but rather require them to be strictly less. Supporting `1` would
+//! entail a more fiddly encoding and is not necessary for our usecase.
```

Comment:
It might be worth documenting the largest represented number, say, for the 16-bit fixed type.